### DISCUSSION TITLE
creation-time of files are changed

### DISCUSSION THREAD

**Discussion by DirkTurlach** [abraunegg/onedrive#2776] [General]
I recognized that the creation-time of files are the time of downloaded to my linux-system. 
That's a big problem for me, because I often looking for file of a specific time, especially in Obsidian.

Is there a possibility to have the original metadata of files (especially creation an modification time)?
And how can the dates of files where already on my system fixed (not with touch, better with new download from Onedrive)

Thanks for help an best regards Dirk

**Comment by abraunegg** [comment]
@DirkTurlach 
Please revirew the RC3 architecture documentation and how v2.5.x operates as timestamp operation has been totally overhauled for the upcomming new release.

**Reply by DirkTurlach** [reply]
Thanks, for the very fast answer.

It ist also correct, that the creation-time is changed in v2.4.x?
In v2.5.x it is changed/fixed?

Would you recommend to alredy v2.5.x for productiv work or should we wait for the stable version?

**Comment by abraunegg** [comment]
@DirkTurlach 
> It ist also correct, that the creation-time is changed in v2.4.x?

I cant say any more what the v2.4.x state is. I have been 18 months into a 100% re-write of v2.5.x

> In v2.5.x it is changed/fixed?

Consider the following file, uploaded online (so will not exist on Linux, yet):

![image](https://github.com/user-attachments/assets/2fc825d6-a9a1-47e1-90c9-10c636ae4314)

This is the JSON that the OneDrive API responds with for this file:
```
{
	"@odata.etag": "\"{6BCBA590-0A8B-427E-9124-D04631D7B5B3},1\"",
	"@odata.type": "#microsoft.graph.driveItem",
	"cTag": "\"c:{6BCBA590-0A8B-427E-9124-D04631D7B5B3},1\"",
	"eTag": "\"{6BCBA590-0A8B-427E-9124-D04631D7B5B3},1\"",
	"file": {
		"hashes": {
			"quickXorHash": "oJRigc8JSjmQwQxhjAMZzBCGOZA="
		},
		"mimeType": "text/plain"
	},
	"fileSystemInfo": {
		"createdDateTime": "2024-08-15T20:05:51Z",
		"lastModifiedDateTime": "2024-08-15T20:05:51Z"
	},
	"id": "01PTBOWVUQUXFWXCYKPZBJCJGQIYY5PNNT",
	"name": "new_online_file.txt",
	"parentReference": {
		"driveId": "b!redacted",
		"driveType": "documentLibrary",
		"id": "01PTBOWVV6Y2GOVW7725BZO354PWSELRRZ",
		"path": "/drives/redacted/root:",
		"siteId": "f0d07a92-98ae-4a22-a2f8-8ae081f196ca"
	},
	"size": 20
}
```

In debug mode you can see that the function is called to set times after the file is downloaded:
```
DEBUG: Calling setTimes() for this file: ./new_online_file.txt
```

When you use `ls -la` on the file this yeilds:
```
-rw-------. 1 alex alex  20 Aug 16 06:05 new_online_file.txt
```

When you use `stat` on the file this yields:
```
[alex@fedora OneDrive]$ stat new_online_file.txt 
  File: new_online_file.txt
  Size: 20              Blocks: 8          IO Block: 4096   regular file
Device: 0,40    Inode: 579274      Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/    alex)   Gid: ( 1000/    alex)
Context: unconfined_u:object_r:user_home_t:s0
Access: 2024-08-16 06:10:10.670827462 +1000
Modify: 2024-08-16 06:05:51.000000000 +1000
Change: 2024-08-16 06:10:10.372834007 +1000
 Birth: 2024-08-16 06:10:10.056840948 +1000
[alex@fedora OneDrive]$ 
```

The method used to do this is: https://dlang.org/phobos/std_file.html#setTimes

There is zero method to change the 'Birth' date within the D Language.

In Linux, the `Birth` or `Creation` time of a file is typically set automatically when the file is created by the filesystem. Unfortunately, most Linux filesystems (like ext4) do not provide a direct way to modify the birth time of a file after it has been created. The `Birth` time is generally immutable and represents the moment when the filesystem inode was created.

> Would you recommend to alredy v2.5.x for productiv work or should we wait for the stable version?

It is probably 3-4 weeks from GM release unless some absolute show stopper is found. It has been run through close to 350 use case tests manually, unknown number of folk already using / testing the RC series.
